File: src/main/java/com/thing2x/smqd/net/telnet/LoginShell.java
package com.thing2x.smqd.net.telnet;

import com.thing2x.smqd.util.StringUtil;
import net.wimpi.telnetd.io.BasicTerminalIO;
import net.wimpi.telnetd.io.toolkit.Editfield;
import net.wimpi.telnetd.net.Connection;
import net.wimpi.telnetd.net.ConnectionEvent;
import net.wimpi.telnetd.shell.Shell;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.EOFException;
import java.io.IOException;

// Created on 2005. 6. 21, Kwon, Yeong Eon

/**
 * This class is an example implmentation of a Shell.<br>
 */
public class LoginShell implements Shell
{
  private static Logger logger = LoggerFactory.getLogger(LoginShell.class);

  private Connection connection;
  private BasicTerminalIO term;

  private String[] allowPattern;
  private LoginShellDelegate delegate;

  public LoginShell()
  {
  }


  public void setDelegate(LoginShellDelegate delegate) { this.delegate = delegate; }
  public void setAllow(String allow)
  {
    allowPattern = StringUtil.split(allow, " , \t\n\r");
  }

  private boolean checkAllow(String address)
  {
    if (allowPattern == null)
      return true;

    for (String pattern : allowPattern)
    {
      if (StringUtil.compareCaseWildExp(address, pattern) == 0)
        return true;
    }

    return false;
  }

  /**
   * Method that runs a shell
   *
   * @param con Connection that runs the shell.
   */
  public void run(Connection con)
  {
    try
    {
      String address = con.getConnectionData().getHostAddress();
      if (!checkAllow(address))
      {
        con.close();
        return;
      }

      connection = con;
      term = connection.getTerminalIO();
      // dont forget to register listener
      connection.addConnectionListener(this);

      // clear the screen and start from zero
      term.eraseScreen();
      term.homeCursor();

      Editfield ef = null;

      for (int i = 0; i < 3; i++)
      {
        term.write("Login: ");
        ef = new Editfield(term, "login", 50);
        ef.run();
        String username = ef.getValue();

        term.write("\r\nPassword: ");
        ef = new Editfield(term, "passwd", 50);
        ef.setPasswordField(true);
        ef.run();
        String password = ef.getValue();
        term.flush();

        if (delegate == null)
        {
          if ("admin".equals(username) && "password".equals(password))
          {
            connection.setNextShell("bsh");
            return;
          }
          else if ("dummy".equals(username) && "dummy".equals(password))
          {
            connection.setNextShell("dummy");
            return;
          }
        }
        else
        {
          if (delegate.login(username, password))
          {
            connection.setNextShell("bsh");
            return;
          }
        }

        term.write("\r\nLogin incorrect\r\n\r\n");
      }

      term.homeCursor();
      term.eraseScreen();
      term.write("Goodbye!.\r\n\r\n");
      term.flush();

      connection.close();
    }
    catch (EOFException e)
    {
      logger.info("Client send quit signal.");
    }
    catch (Exception ex)
    {
      logger.error("run()", ex);
    }
  }

  // this implements the ConnectionListener!
  public void connectionTimedOut(ConnectionEvent ce)
  {
    try
    {
      term.write("CONNECTION_TIMEDOUT");
      term.flush();
      // close connection
      connection.close();
    }
    catch (Exception ex)
    {
      logger.error("connectionTimedOut()", ex);
    }
  }

  public void connectionIdle(ConnectionEvent ce)
  {
    try
    {
      term.write("CONNECTION_IDLE");
      term.flush();
    }
    catch (IOException e)
    {
      logger.error("connectionIdle()", e);
    }

  }

  public void connectionLogoutRequest(ConnectionEvent ce)
  {
    try
    {
      term.write("CONNECTION_LOGOUTREQUEST");
      term.flush();
    }
    catch (Exception ex)
    {
      logger.error("connectionLogoutRequest()", ex);
    }
  }

  public void connectionSentBreak(ConnectionEvent ce)
  {
    try
    {
      term.write("CONNECTION_BREAK");
      term.flush();
    }
    catch (Exception ex)
    {
      logger.error("connectionSentBreak()", ex);
    }
  }

  public static Shell createShell()
  {
    return new LoginShell();
  }
}

File: src/main/scala/com/thing2x/smqd/net/telnet/LoginShell.scala
// Copyright 2018 UANGEL
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.thing2x.smqd.net.telnet

import java.io.{EOFException, IOException}
import java.net.InetAddress

import com.typesafe.scalalogging.StrictLogging
import net.wimpi.telnetd.io.BasicTerminalIO
import net.wimpi.telnetd.io.toolkit.Editfield
import net.wimpi.telnetd.net.{Connection, ConnectionEvent}
import net.wimpi.telnetd.shell.Shell

// 10/15/18 - Created by Kwon, Yeong Eon

object LoginShell {

  trait Delegate {
    def login(shell: LoginShell, user: String, password: String): Boolean = true
    def allow(shell: LoginShell, remoteAddress: InetAddress): Boolean = {
      //val address = remoteAddress.getHostName
      //StringUtil.compareCaseWildExp(address, pattern) == 0
      true
    }
  }

  def setDelegate(delegate: LoginShell.Delegate): Unit = {
    this.delegate = Option(delegate)
  }

  private var delegate: Option[LoginShell.Delegate] = None

  def createShell: Shell = new LoginShell(delegate)
}

/**
  *
  */
class LoginShell(delegate: Option[LoginShell.Delegate]) extends Shell with StrictLogging {

  private var term: BasicTerminalIO = _
  private var connection: Connection = _

  override def run(conn: Connection): Unit = {
    try {
      val address = conn.getConnectionData.getInetAddress
      if (!(delegate.isDefined && delegate.get.allow(this, address))) {
        conn.close()
        return
      }

      connection = conn
      term = connection.getTerminalIO

      // dont forget to register listener
      connection.addConnectionListener(this)

      // clear the screen and start from zero
      term.eraseScreen()
      term.homeCursor()

      // allow retry 3 times
      val logined = 1 to 3 find { _ =>
        term.write("Login: ")
        var ef = new Editfield(term, "login", 50)
        ef.run()
        val username = ef.getValue

        term.write("\r\nPassword: ")
        ef = new Editfield(term, "passwd", 50)
        ef.setPasswordField(true)
        ef.run()
        val password = ef.getValue
        term.flush()

        if ("dummy" == username && "dummy" == password) {
          connection.setNextShell("dummy")
          true
        }
        else if (delegate.isEmpty && "admin" == username && "password" == password) {
          connection.setNextShell("bsh")
          true
        }
        else if (delegate.isDefined && delegate.get.login(this, username, password)) {
          connection.setNextShell("bsh")
          true
        }
        else {
          term.write("\r\n")
          false
        }
      }

      connection.removeConnectionListener(this)

      if (logined.isEmpty) {
        term.write("\r\nLogin incorrect\r\n\r\n")

        term.homeCursor()
        term.eraseScreen()
        term.write("Goodbye!.\r\n\r\n")
        term.flush()

        connection.close()
      }
    } catch {
      case _: EOFException =>
        logger.info("Client send quit signal.")
      case ex: Exception =>
        logger.error("LoginShell", ex.getMessage)
    }
  }

  override def connectionIdle(connectionEvent: ConnectionEvent): Unit = try {
    term.write("CONNECTION_IDLE")
    term.flush()
  } catch {
    case e: IOException =>
      logger.error("connectionIdle()", e)
  }

  override def connectionTimedOut(connectionEvent: ConnectionEvent): Unit = try {
    term.write("CONNECTION_TIMEDOUT")
    term.flush()
    // close connection
    connection.close()
  } catch {
    case ex: Exception =>
      logger.error("connectionTimedOut()", ex)
  }


  override def connectionLogoutRequest(connectionEvent: ConnectionEvent): Unit = try {
    term.write("CONNECTION_LOGOUTREQUEST")
    term.flush()
  } catch {
    case ex: Exception =>
      logger.error("connectionLogoutRequest()", ex)
  }

  override def connectionSentBreak(connectionEvent: ConnectionEvent): Unit = try {
    term.write("CONNECTION_BREAK")
    term.flush()
  } catch {
    case ex: Exception =>
      logger.error("connectionSentBreak()", ex)
  }
}
